src/service.py

```diff
@@ -0,0 +1,121 @@
+        fn = func.__name__  # This should be a verb
+        if not self.snap_client:
+            logger.error(
+                "cannot %s %s because it's unable to get snap client", fn, self.service_name
+            )
+            return None
+        logger.info("%s %s", fn, self.service_name)
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+def guard_installed(func: Callable) -> Callable:
+    """Ensure the snap is installed before running a snap operation."""
+
+    @wraps(func)
+    def wrapper(self: "SnapService", *args: Any, **kwargs: dict[Any, Any]) -> Optional[Any]:
+        fn = func.__name__  # This should be a verb
+        if not self.check_installed():
+            logger.error("cannot %s %s because it's not installed", fn, self.service_name)
+            return None
+        logger.info("%s %s", fn, self.service_name)
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class SnapService:
+    """A class representing the a snap service (only support one service)."""
+
+    def __init__(self, name: str, service_name: str) -> None:
+        """Initialize the class."""
+        self.name = name
+        self.service_name = service_name
+        self.snap_client = self.get_client()
+
+    def get_client(self) -> Optional[snap.Snap]:
+        """Return the snap client, or None if the snap cannot be found."""
+        try:
+            client = snap.SnapCache()[self.name]
+        except snap.SnapNotFoundError as e:
+            logger.warning("unable to get snap client: %s", str(e))
+            return None
+        return client
+
+    def install(self, resource: str) -> None:
+        """Install or refresh the snap."""
+        try:
+            logger.debug("installing snap.")
+            logger.debug("fetching from resource.")
+            snap_client = snap.install_local(resource, dangerous=True)
+        except snap.SnapError as e:
+            logger.error("failed to install snap: %s", str(e))
+            raise e  # need to crash on_install event if it's not okay
+        else:
+            self.snap_client = snap_client
+            logger.info("installed %s snap.", self.name)
+
+    @guard_client
+    @guard_installed
+    def remove(self) -> None:
+        """Remove the snap."""
+        try:
+            snap.remove([self.name])
+        except snap.SnapError as e:
+            logger.info("failed to remove snap.")
+            logger.error(str(e))
+        else:
+            logger.info("removed %s snap.", self.name)
+
+    @guard_client
+    def check_active(self) -> bool:
+        """Return True if the snap service is active."""
+        return self.snap_client.services.get(self.service_name, {}).get("active", False)  # type: ignore
+
+    @guard_client
+    def check_installed(self) -> bool:
+        """Return True if the snap is installed."""
+        return self.snap_client.present  # type: ignore
```

Comment:
Yep, that make sense, thanks!

Reply:
> if not snap:
  add_status(Blocked("snap is not installed yet"))
should be not necessary in collect_status hook since the install process should be one-off during install hook, and probably best to use with `defer` in configure hook

Reply:
hmm.. I guess we need it, user can remove it themselves for any reason (via `juju ssh`), without it the status in juju is not correct

Reply:
Updated